ASoC: Intel: soc-acpi-intel-arl-match: Add rt722 and rt1320 support

macchian opened the pull request:
This patch adds support for the rt722 multi-function codec and the rt1320 amplifier in the ARL board configuration.

Link 0: RT722 codec with three endpoints: Headset, Speaker, and DMIC.
Link 2: RT1320 amplifier.

Note:
The Speaker endpoint on the RT722 codec is not used.

macchian committed:
ASoC: Intel: soc-acpi-intel-arl-match: Add rt722 and rt1320 support

This patch adds support for the rt722 multi-function codec and the
rt1320 amplifier in the ARL board configuration.

Link 0: RT722 codec with three endpoints: Headset, Speaker, and DMIC.
Link 2: RT1320 amplifier.

Note:
The Speaker endpoint on the RT722 codec is not used.

Signed-off-by: Mac Chiang <mac.chiang@intel.com>

bardliao commented on sound/soc/intel/common/soc-acpi-intel-arl-match.c:
```diff
@@ -44,6 +44,31 @@ static const struct snd_soc_acpi_endpoint spk_3_endpoint = {
+		.aggregated = 0,
+		.group_position = 0,
+		.group_id = 0,
+	},
```
We need to remove the Speaker endpoint if it is not used.

plbossart replied:
yeah, but there's no way to know is the speaker on the RT1320 is the only one in the system, or if it's used as a companion to that of the RT722.

The notion of endpoint was already not super useful and mostly introduced as a way to predict the future with audio composition, but it's already obsolete in that endpoints are now optional.... We'd need to check if the functions are described in ACPI for sure.

bardliao replied:
@plbossart Can you take a look at https://github.com/thesofproject/linux/pull/5188?

bardliao replied:
@charleskeepax has a good suggestion on https://github.com/thesofproject/linux/pull/5186 that can make a DAI be optional and present by default.

macchian replied:
> @charleskeepax has a good suggestion on #5186 that can make a DAI be optional and present by default.

@bardliao, thank you for your advice. By referencing @charleskeepax #5186 suggestion on cs42l43, I could apply the quirk_exclude codes into RT722 as a short-term once #5186 merged. In the end, #5188 implementing the "endpoints present by sdca acpi function" will be a final solution afterwards.

macchian committed:
ASoC: sdw_utils: Add quirk to exclude amplifier function

When SKUs use the multi-function codec, which integrates
Headset, Amplifier and DMIC. The corresponding quirks provide
options to support internal amplifier/DMIC or not.

In the case of RT722, this SKU excludes the internal amplifier and
use an additional amplifier instead.

Signed-off-by: Mac Chiang <mac.chiang@intel.com>

macchian committed:
ASoC: intel: sof_sdw: add quirk for Dell SKU

This patch adds a quirk to include the codec amplifier
function for this Dell SKU.

Note: In this SKU '0CF0', the RT722 codec amplifier is
excluded, and an external amplifier is used instead.

Signed-off-by: Mac Chiang <mac.chiang@intel.com>

bardliao commented:
LGTM. But the commits should re-order. [ASoC: sdw_utils: Add quirk to exclude amplifier function](https://github.com/thesofproject/linux/pull/5187/commits/0fde35b3322d28ecb61be331473508110d5cf6f6) should be the first commit. Also, please mention rt721 in the commit message.

macchian commented:
> LGTM. But the commits should re-order. [ASoC: sdw_utils: Add quirk to exclude amplifier function](https://github.com/thesofproject/linux/pull/5187/commits/0fde35b3322d28ecb61be331473508110d5cf6f6) should be the first commit. Also, please mention rt721 in the commit message.

@bardliao ,thanks for your comments. Done.

bardliao commented on sound/soc/sdw_utils/soc_sdw_utils.c:
```diff
@@ -363,6 +363,8 @@ struct asoc_sdw_codec_info codec_info_list[] = {
 				.widgets = generic_spk_widgets,
 				.num_widgets = ARRAY_SIZE(generic_spk_widgets),
+				.quirk = SOC_SDW_CODEC_SPKR,
+				.quirk_exclude = true,
```
The quirk is added to the rt721 dai. Is it correct?

macchian replied:
> The quirk is added to the rt721 dai. Is it correct?

Oh, thanks for pointing out, resolved.

macchian commented:
These PRs has been verified on board. We can merge now.
